File: solutions/leetcode/922-sort_array_by_parity-ii.java
class Solution {
    public int[] sortArrayByParityII(int[] nums) {
        int even = 0;
        int odd = 1;
        int n = nums.length;

        while (even < n && odd < n) {
            if (nums[even] % 2 == 0) {
                even += 2;
            }
            else if (nums[odd] % 2 == 1) {
                odd += 2;
            }
            else {
                swap(nums, even, odd);
                even += 2;
                odd += 2;
            }
        }

        return nums;
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: solutions/leetcode/922.sort-array-by-parity-ii.cpp
/*
 * @lc app=leetcode id=922 lang=cpp
 *
 * [922] Sort Array By Parity II
 */

// @lc code=start
class Solution {
public:
    vector<int> sortArrayByParityII(vector<int>& nums) {
        int even = 0;
        int odd = 1;

        while (even < nums.size() && odd < nums.size()) {
            if (nums[even] % 2 == 0) {
                even += 2;
            }
            else if (nums[odd] % 2 != 0) {
                odd += 2;
            }
            else {
                std::swap(nums[even], nums[odd]);
                even += 2;
                odd += 2;
            }
        }

        return nums;
    }
};
// @lc code=end

